TestPipelineBuildLifecycleGo flake

TestPipelineBuildLifecycleGo failed in https://travis-ci.com/github/pachyderm/pachyderm/jobs/399167346

```
=== RUN   TestPipelineBuildLifecycleGo
../../../../etc/testing/pipeline-build/input/111.txt: 3 B / 3 B  100.00% ? p/s 0
../../../../etc/testing/pipeline-build/input/1.txt: 1 B / 1 B [ 100.00% ? p/s 0s
../../../../etc/testing/pipeline-build/input/11.txt: 2 B / 2 B  100.00% ? p/s 0s
go.mod: 131 B / 131 B [=======================================] 100.00% ? p/s 0s
go.sum: 169 B / 169 B [=======================================] 100.00% ? p/s 0s
main.go: 795 B / 795 B [======================================] 100.00% ? p/s 0s
file /1.txt not found in repo test-pipeline-build at commit d4ec0aec17fa49248db2a5d388e5bfc1
failed to find
  "0001"
in the text:
  """
  """
```

It looks like sometimes the test fails to load the source code into the `source` branch:
```
file /1.txt not found in repo test-pipeline-build at commit a010b8bfd21d4f5280628e0de454c3e0
failed to find
  "___1"
in the text:
  """

  """
--- FAIL: TestPipelineBuildLifecycleGo (46.42s)
    cmds_test.go:630: No error is expected but got exit status 1
    cmds_test.go:630: current stack:
        goroutine 14 [running]:
        runtime/debug.Stack(0x25a47a0, 0xc000480780, 0xc00038bd98)
                /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/pachyderm/pachyderm/src/client/pkg/require.fatal(0x262e8c0, 0xc0005db500, 0x0, 0x0, 0x0, 0x2252bca, 0x1f, 0xc000209a10, 0x1, 0x1)
                /home/mjs/clients/s3g_to_kubeflow_20191212/src/github.com/pachyderm/pachyderm/src/client/pkg/require/require.go:534 +0xd9
        github.com/pachyderm/pachyderm/src/client/pkg/require.NoError(0x262e8c0, 0xc0005db500, 0x25a47a0, 0xc000480780, 0x0, 0x0, 0x0)
                /home/mjs/clients/s3g_to_kubeflow_20191212/src/github.com/pachyderm/pachyderm/src/client/pkg/require/require.go:394 +0xef
        github.com/pachyderm/pachyderm/src/server/pps/cmds.verifyPipelineBuildOutput(0xc0005db500, 0x2221841, 0x1)
                /home/mjs/clients/s3g_to_kubeflow_20191212/src/github.com/pachyderm/pachyderm/src/server/pps/cmds/cmds_test.go:733 +0xdf
        github.com/pachyderm/pachyderm/src/server/pps/cmds.testPipelineBuildLifecycle(0xc0005db500, 0x2221a70, 0x2, 0x2221a70, 0x2)
                /home/mjs/clients/s3g_to_kubeflow_20191212/src/github.com/pachyderm/pachyderm/src/server/pps/cmds/cmds_test.go:727 +0x508
        github.com/pachyderm/pachyderm/src/server/pps/cmds.TestPipelineBuildLifecycleGo(0xc0005db500)
                /home/mjs/clients/s3g_to_kubeflow_20191212/src/github.com/pachyderm/pachyderm/src/server/pps/cmds/cmds_test.go:630 +0x4e
        testing.tRunner(0xc0005db500, 0x22f8708)
                /usr/local/go/src/testing/testing.go:909 +0xc9
        created by testing.(*T).Run
                /usr/local/go/src/testing/testing.go:960 +0x350
FAIL
FAIL    github.com/pachyderm/pachyderm/src/server/pps/cmds      46.464s
FAIL
$ pachctl list pipeline
NAME                      VERSION INPUT                                                               CREATED        STATE / LAST JOB  DESCRIPTION
test-pipeline-build       3       (test-pipeline-build_build:/ ⨯ in:/* ⨯ test-pipeline-build_build:/) 16 seconds ago running / failure
test-pipeline-build_build 3       test-pipeline-build_build:/                                         16 seconds ago running / failure
$ pachctl list job
ID                               PIPELINE                  STARTED                DURATION           RESTART PROGRESS  DL UL STATE
6df32a51009248029a9925b560a39c38 test-pipeline-build       Less than a second ago Less than a second 0       0 + 0 / 0 0B 0B failure: inputs failed: build
b10b9fea053c4814a8d9712dccd3d636 test-pipeline-build_build Less than a second ago Less than a second 0       0 + 0 / 1 0B 0B failure: datum failed
b181f93528364f58ac415becb9bdcbe0 test-pipeline-build_build 1 second ago           Less than a second 0       0 + 1 / 1 0B 0B success

$ pc inspect job b10b9fea053c4814a8d9712dccd3d636
ID: b10b9fea053c4814a8d9712dccd3d636
Pipeline: test-pipeline-build_build
Started: 23 seconds ago
Duration: Less than a second
State: failure
Reason: datum failed
Processed: 0
Failed: 1
Skipped: 0
Recovered: 0
Total: 1
Data Downloaded: 0B
Data Uploaded: 0B
Download Time: Less than a second
Process Time: Less than a second
Upload Time: Less than a second
Datum Timeout: (duration: nil Duration)
Job Timeout: (duration: nil Duration)
Worker Status:
WORKER              JOB                 DATUM               STARTED             QUEUE
Restarts: 0
ParallelismSpec: constant:1
Input:
{
  "pfs": {
    "name": "source",
    "repo": "test-pipeline-build_build",
    "branch": "source",
    "commit": "7dce3c5111734ba8a585f3534e63a5e1",
    "glob": "/"
  }
}

Transform:
{
  "image": "pachyderm/go-build:1.12.0-a531c0d4b8e38271d465e41f8687eb055486d82f",
  "cmd": [
    "sh",
    "./build.sh"
  ]
}
Output Commit: c3310f7047c64575bb628269e11861da
$ pc list datum --job=b10b9fea053c4814a8d9712dccd3d636
Error: unknown flag: --job
Usage:
  pachctl list datum <job> [flags]

Flags:
  -h, --help            help for datum
  -o, --output string   Output format when --raw is set: "json" or "yaml" (default "json")
      --page int        Specify the page of results to send
      --pageSize int    Specify the number of results sent back in a single page
      --raw             Disable pretty printing; serialize data structures to an encoding such as json or yaml

Global Flags:
      --no-color   Turn off colors.
  -v, --verbose    Output verbose logs

unknown flag: --job
$ pc list datum b10b9fea053c4814a8d9712dccd3d636
ID                                                                   STATUS TIME
6e16f09447614c8034c2495eb6c8cc9c8f4673700c1d329e02ef7fb785900e38ea61 -      -
$ pc inspect datum b10b9fea053c4814a8d9712dccd3d636@6e16f09447614c8034c2495eb6c8cc9c8f4673700c1d329e02ef7fb785900e38ea61
expected 2 arguments, got 1

Usage:
  pachctl inspect datum <job> <datum> [flags]

Flags:
  -h, --help            help for datum
  -o, --output string   Output format when --raw is set: "json" or "yaml" (default "json")
      --raw             Disable pretty printing; serialize data structures to an encoding such as json or yaml

Global Flags:
      --no-color   Turn off colors.
  -v, --verbose    Output verbose logs

$ pc logs --job=b10b9fea053c4814a8d9712dccd3d636
go version go1.15.3 linux/amd64
package .: no Go files in /pfs/source
go version go1.15.3 linux/amd64
package .: no Go files in /pfs/source
go version go1.15.3 linux/amd64
package .: no Go files in /pfs/source

$ pc list commit test-pipeline-build_build
REPO                      BRANCH COMMIT                           FINISHED      SIZE     PROGRESS DESCRIPTION
test-pipeline-build_build build  c3310f7047c64575bb628269e11861da 3 minutes ago 0B       ▇▇▇▇▇▇▇▇
test-pipeline-build_build source 7dce3c5111734ba8a585f3534e63a5e1 3 minutes ago 0B       ▇▇▇▇▇▇▇▇
test-pipeline-build_build build  dadc0ff8a6894a48bf066de29bee1f19 3 minutes ago 2.137MiB ▇▇▇▇▇▇▇▇
test-pipeline-build_build build  4874e67e1c8c44ca8b54df1cd686aa3d 3 minutes ago 2.137MiB ▇▇▇▇▇▇▇▇
test-pipeline-build_build source 5c1273cbc93742c4b19d112ccfc716b6 3 minutes ago 1.069KiB ▇▇▇▇▇▇▇▇
test-pipeline-build_build build  b05a688b075e4dd5bad6b7e5cc1f413a 3 minutes ago 2.137MiB ▇▇▇▇▇▇▇▇
test-pipeline-build_build build  60fcfeead1344ff8894dd3e88ac09037 3 minutes ago 2.137MiB ▇▇▇▇▇▇▇▇
test-pipeline-build_build source d93c863795d943779f164b02eb04b5ee 4 minutes ago 1.069KiB ▇▇▇▇▇▇▇▇

$ pc list file test-pipeline-build_build@source
NAME TYPE SIZE

$ pc list file test-pipeline-build_build@source~1
NAME     TYPE SIZE
/go.mod  file 131B
/go.sum  file 169B
/main.go file 795B
```

--- a/etc/kube/start-minikube-vm.sh
+++ b/etc/kube/start-minikube-vm.sh
@@ -25,7 +25,7 @@ function get_images {
 export -f get_images
 
 function start_minikube {
-    minikube start ${MINIKUBE_FLAGS[@]}
+    minikube start "${MINIKUBE_FLAGS[@]}"
 }
 export -f start_minikube
 

--- a/etc/testing/lint.sh
+++ b/etc/testing/lint.sh
@@ -5,24 +5,23 @@ set -e
 GIT_REPO_DIR=$(cd "$( dirname "${BASH_SOURCE[0]}" )" 2>&1 > /dev/null && git rev-parse --show-toplevel)
 cd "${GIT_REPO_DIR}"
 
-# Collect untracked args into '\( -path abc -o -path def -o ... \) -prune'
+# Collect untracked args into '\( -o -path abc -o -path def -o ... \) -prune'
 # argument to be passed to 'find'
 skip_paths=()
 for file in $(git status --porcelain | grep '^??' | sed 's/^?? //'); do
+  # Convert to relative path, as required by -path
   file="./$(realpath --relative-to="." "${file}")"
-  if [[ "${#skip_paths[@]}" -gt 0 ]]; then
-    skip_paths+=( -o )
-  fi
-  skip_paths+=( -path "${file%/}" ) # strip trailing slash as req'd by -path
+  # 1. Always prefix with -o (i.e. don't fencepost; find always has an argument
+  #    before skip_paths, so including the initial -o works whether skip_paths
+  #    is empty or not
+  # 2. strip trailing slash as required by -path
+  skip_paths+=( -o -path "${file%/}" )
 done
 
 go get -u golang.org/x/lint/golint
 find "./src" \
-  \( -path "*.pb.go" -o "${skip_paths[@]}" \) -prune -o -name '*.go' -print \
+  \( -path "*.pb.go" -o -path "*pkg/tar*" "${skip_paths[@]}" \) -prune -o -name '*.go' -print \
 | while read -r file; do
-    if [[ "${file}" == *pkg/tar* ]]; then
-        continue
-    fi
     golint -set_exit_status "$file";
 done;
 
@@ -38,7 +37,7 @@ staticcheck "${GIT_REPO_DIR}/..."
 
 # shellcheck disable=SC2046
 find . \
-  \( -path ./etc/plugin -o "${skip_paths[@]}" \) -prune -o -name "*.sh" -print \
+  \( -path ./etc/plugin "${skip_paths[@]}" \) -prune -o -name "*.sh" -print \
 | while read -r file; do
     shellcheck -e SC1091 -e SC2010 -e SC2181 -e SC2004 -e SC2219 "${file}"
 done

--- a/src/server/pfs/server/driver.go
+++ b/src/server/pfs/server/driver.go
@@ -3050,10 +3050,11 @@ func (d *driver) putFiles(pachClient *client.APIClient, s *putFileServer) error 
 		var records *pfs.PutFileRecords
 		logPutFileStart(req)
 		defer func() { logPutFileEnd(req, start, records, retErr) }() //late binding
-		records, retErr = d.putFile(pachClient, req.File, req.Delimiter, req.TargetFileDatums,
+		var err error
+		records, err = d.putFile(pachClient, req.File, req.Delimiter, req.TargetFileDatums,
 			req.TargetFileBytes, req.HeaderRecords, req.OverwriteIndex, req.Delete, r)
-		if retErr != nil {
-			return retErr
+		if err != nil {
+			return err
 		}
 		mu.Lock()
 		defer mu.Unlock()

--- a/src/server/pfs/server/driver.go
+++ b/src/server/pfs/server/driver.go
@@ -4666,6 +4666,7 @@ func (s *putFileServer) Peek() (*pfs.PutFileRequest, error) {
 
 func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_PutFileServer, f func(*pfs.PutFileRequest, io.Reader) error) (oneOff bool, repo string, branch string, retErr error) {
 
+	var pl = newPathLock(d.putFileLimiter)
 	var pr *io.PipeReader
 	var pw *io.PipeWriter
 	var req *pfs.PutFileRequest
@@ -4755,18 +4756,23 @@ func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_Put
 				case "http":
 					fallthrough
 				case "https":
-					d.putFileLimiter.Acquire()
 					resp, err := http.Get(req.Url)
 					if err != nil {
 						return false, "", "", err
 					} else if resp.StatusCode >= 400 {
 						return false, "", "", errors.Errorf("error retrieving content from %q: %s", req.Url, resp.Status)
 					}
+					if err := pl.start(req.File.Path); err != nil {
+						resp.Body.Close() // drop error as we're failing anyway
+						return false, "", "", errors.Wrapf(err, "could not lock path %q", req.File.Path)
+					}
 					eg.Go(func() (retErr error) {
-						defer d.putFileLimiter.Release()
 						defer func() {
 							if err := resp.Body.Close(); err != nil && retErr == nil {
 								retErr = err
+							}
+							if err := pl.finish(req.File.Path); err != nil && retErr == nil {
+								retErr = errors.Wrapf(err, "could not unlock path %q", req.File.Path)
 							}
 						}()
 						return f(req, resp.Body)
@@ -4791,16 +4797,21 @@ func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_Put
 							}
 							req := *req // copy req so we can make changes
 							req.File = client.NewFile(req.File.Commit.Repo.Name, req.File.Commit.ID, filepath.Join(req.File.Path, strings.TrimPrefix(name, path)))
-							d.putFileLimiter.Acquire()
 							r, err := objClient.Reader(server.Context(), name, 0, 0)
 							if err != nil {
 								return err
 							}
+							if err := pl.start(req.File.Path); err != nil {
+								r.Close() // drop error as we're failing anyway
+								return errors.Wrapf(err, "could not lock path %q", req.File.Path)
+							}
 							eg.Go(func() (retErr error) {
-								defer d.putFileLimiter.Release()
 								defer func() {
 									if err := r.Close(); err != nil && retErr == nil {
 										retErr = err
+									}
+									if err := pl.finish(req.File.Path); err != nil && retErr == nil {
+										retErr = errors.Wrapf(err, "could not unlock path %q", req.File.Path)
 									}
 								}()
 								return f(&req, r)
@@ -4810,16 +4821,21 @@ func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_Put
 							return false, "", "", err
 						}
 					} else {
-						d.putFileLimiter.Acquire()
 						r, err := objClient.Reader(server.Context(), url.Object, 0, 0)
 						if err != nil {
 							return false, "", "", err
 						}
+						if err := pl.start(req.File.Path); err != nil {
+							r.Close() // drop error as we're failing anyway
+							return false, "", "", errors.Wrapf(err, "could not lock path %q", req.File.Path)
+						}
 						eg.Go(func() (retErr error) {
-							defer d.putFileLimiter.Release()
 							defer func() {
 								if err := r.Close(); err != nil && retErr == nil {
 									retErr = err
+								}
+								if err := pl.finish(req.File.Path); err != nil && retErr == nil {
+									retErr = errors.Wrapf(err, "could not unlock path %q", req.File.Path)
 								}
 							}()
 							return f(req, r)
@@ -4833,18 +4849,30 @@ func (d *driver) forEachPutFile(pachClient *client.APIClient, server pfs.API_Put
 				pw.Close() // can't error
 			}
 			if req.Delete {
-				d.putFileLimiter.Acquire()
-				eg.Go(func() error {
-					defer d.putFileLimiter.Release()
+				if err := pl.start(req.File.Path); err != nil {
+					return false, "", "", errors.Wrapf(err, "could not lock path %q", req.File.Path)
+				}
+				eg.Go(func() (retErr error) {
+					defer func() {
+						if err := pl.finish(req.File.Path); err != nil && retErr == nil {
+							retErr = errors.Wrapf(err, "could not unlock path %q", req.File.Path)
+						}
+					}()
 					return f(req, nil)
 				})
 				continue
 			}
 			pr, pw = io.Pipe()
 			pr := pr
-			d.putFileLimiter.Acquire()
-			eg.Go(func() error {
-				defer d.putFileLimiter.Release()
+			if err := pl.start(req.File.Path); err != nil {
+				return false, "", "", errors.Wrapf(err, "could not lock path %q", req.File.Path)
+			}
+			eg.Go(func() (retErr error) {
+				defer func() {
+					if err := pl.finish(req.File.Path); err != nil && retErr == nil {
+						retErr = errors.Wrapf(err, "could not unlock path %q", req.File.Path)
+					}
+				}()
 				if err := f(req, pr); err != nil {
 					// needed so the parent goroutine doesn't block
 					pr.CloseWithError(err)

--- a/src/server/pfs/server/driver.go
+++ b/src/server/pfs/server/driver.go
@@ -3034,7 +3034,7 @@ func logPutFileEnd(req *pfs.PutFileRequest, start time.Time, records *pfs.PutFil
 			bytes += float64(r.SizeBytes)
 		}
 	}
-	dur := time.Now().Sub(start).Truncate(time.Millisecond)
+	dur := time.Since(start).Truncate(time.Millisecond)
 	rate := float64(bytes) / dur.Seconds()
 	log.Debugf(`pfs.API.PutFile/%s {"to":%q,"sz":%q,"dur":"%s (%s/s)","err":"%v"`,
 		verb, pretty.CompactPrintFile(req.File), units.BytesSize(bytes), dur, units.BytesSize(rate), retErr)

--- a/src/server/pfs/server/driver.go
+++ b/src/server/pfs/server/driver.go
@@ -3029,8 +3029,10 @@ func logPutFileEnd(req *pfs.PutFileRequest, start time.Time, records *pfs.PutFil
 		verb = "Deleted"
 	}
 	var bytes float64
-	for _, r := range records.Records {
-		bytes += float64(r.SizeBytes)
+	if records != nil {
+		for _, r := range records.Records {
+			bytes += float64(r.SizeBytes)
+		}
 	}
 	dur := time.Now().Sub(start).Truncate(time.Millisecond)
 	rate := float64(bytes) / dur.Seconds()

--- a/src/server/pfs/server/driver.go
+++ b/src/server/pfs/server/driver.go
@@ -28,6 +28,7 @@ import (
 	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
 	"github.com/pachyderm/pachyderm/src/client/pkg/grpcutil"
 	pfsserver "github.com/pachyderm/pachyderm/src/server/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pfs/pretty"
 	"github.com/pachyderm/pachyderm/src/server/pkg/ancestry"
 	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
 	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
@@ -43,6 +44,7 @@ import (
 	"github.com/pachyderm/pachyderm/src/server/pkg/watch"
 
 	etcd "github.com/coreos/etcd/clientv3"
+	units "github.com/docker/go-units"
 	"github.com/gogo/protobuf/proto"
 	"github.com/gogo/protobuf/types"
 	log "github.com/sirupsen/logrus"
@@ -3013,17 +3015,43 @@ func (d *driver) scratchFilePrefix(file *pfs.File) (string, error) {
 	return path.Join(d.scratchCommitPrefix(file.Commit), cleanedPath), nil
 }
 
+func logPutFileStart(req *pfs.PutFileRequest) {
+	verb := "Writing"
+	if req.Delete {
+		verb = "Deleting"
+	}
+	log.Debugf(`pfs.API.PutFile/%s {"to":%q}`, verb, pretty.CompactPrintFile(req.File))
+}
+
+func logPutFileEnd(req *pfs.PutFileRequest, start time.Time, records *pfs.PutFileRecords, retErr error) {
+	verb := "Wrote"
+	if req.Delete {
+		verb = "Deleted"
+	}
+	var bytes float64
+	for _, r := range records.Records {
+		bytes += float64(r.SizeBytes)
+	}
+	dur := time.Now().Sub(start).Truncate(time.Millisecond)
+	rate := float64(bytes) / dur.Seconds()
+	log.Debugf(`pfs.API.PutFile/%s {"to":%q,"sz":%q,"dur":"%s (%s/s)","err":"%v"`,
+		verb, pretty.CompactPrintFile(req.File), units.BytesSize(bytes), dur, units.BytesSize(rate), retErr)
+}
+
 func (d *driver) putFiles(pachClient *client.APIClient, s *putFileServer) error {
 	var files []*pfs.File
 	var putFilePaths []string
 	var putFileRecords []*pfs.PutFileRecords
 	var mu sync.Mutex
-	oneOff, repo, branch, err := d.forEachPutFile(pachClient, s, func(req *pfs.PutFileRequest, r io.Reader) error {
-		log.Debugf("Writing to %v@%v:/%v", req.File.Commit.Repo.Name, req.File.Commit.ID, req.File.Path)
-		records, err := d.putFile(pachClient, req.File, req.Delimiter, req.TargetFileDatums,
+	oneOff, repo, branch, err := d.forEachPutFile(pachClient, s, func(req *pfs.PutFileRequest, r io.Reader) (retErr error) {
+		start := time.Now()
+		var records *pfs.PutFileRecords
+		logPutFileStart(req)
+		defer func() { logPutFileEnd(req, start, records, retErr) }() //late binding
+		records, retErr = d.putFile(pachClient, req.File, req.Delimiter, req.TargetFileDatums,
 			req.TargetFileBytes, req.HeaderRecords, req.OverwriteIndex, req.Delete, r)
-		if err != nil {
-			return err
+		if retErr != nil {
+			return retErr
 		}
 		mu.Lock()
 		defer mu.Unlock()

--- a/src/server/pfs/server/pathlock.go
+++ b/src/server/pfs/server/pathlock.go
@@ -54,9 +54,6 @@ func newPathLock(putFileLimiter limit.ConcurrencyLimiter) *pathlock {
 
 func (p *pathlock) start(path string) (retErr error) {
 	// add leading slash, so '/' collides w/ everything.
-	// TODO(msteffen): would paths.go (in this package) or s/s/pkg/paths be
-	// better? The latter doesn't exist in 1.11.x, and the former adds a leading
-	// slash but doesn't clean the path.
 	path = gopath.Clean("/" + path)
 	defer func() {
 		if retErr == nil && p.putFileLimiter != nil {

--- a/src/server/pfs/server/pathlock.go
+++ b/src/server/pfs/server/pathlock.go
@@ -37,7 +37,7 @@ type pathlock struct {
 	// it may proceed
 	blocking string
 
-	// This is used to limit the total concurrency of putFile request. It's
+	// This is used to limit the total # of concurrent of putFile requests. It's
 	// included in pathlock as a convenience, as everywhere that putFile calls
 	// pathlock.start(), it would also call putFileLimiter.Acquire() (and likewise
 	// for pathlock.finish() and putFileLimiter.Release). This is unset for tests.
@@ -65,28 +65,22 @@ func (p *pathlock) start(path string) (retErr error) {
 	}()
 	p.Lock()
 	defer p.Unlock()
-	for { // Check block condition in loop--see sync.Cond documentation
+	for { // Check block condition in a loop--see sync.Cond documentation
 		if blocking := func() string {
-			// Find if there's any path P in 'running' such that 'path' is a prefix of
-			// P
-			//
-			// minGe is the smallest entry in 'running' that is greater than or equal
-			// to 'path'. If there is a P in running such that 'path' is a prefix of
-			// P, then 'path' must be a prefix of 'running[minGe]'[1]
-			minGe := sort.SearchStrings(p.running, path)
-			if minGe < len(p.running) && strings.HasPrefix(p.running[minGe], path) {
-				return p.running[minGe]
+			// Find if there's any P in 'running' such that 'path' is a prefix of P[1]
+			lub := sort.SearchStrings(p.running, path)
+			if lub < len(p.running) && strings.HasPrefix(p.running[lub], path) {
+				return p.running[lub]
 			}
-			// Find if there's any path P in 'running' such that P is a prefix of
-			// 'path'. Do this by checking if every prefix of 'path' is in 'running'.
-			for i := len(path) - 1; i > 0 && minGe > 0; i-- {
-				// Loop invariant: minGe is the smallest entry in running that is >=
-				//   path[:i+1] (note loop starts at len-1).
-				//   path[:i] < path[:i+1] --> path[:i] must be in [0,minGe)
-				// Loop invariant: path[:i] and [0, minGe) are non-empty
-				minGe = sort.SearchStrings(p.running[:minGe], path[:i])
-				if minGe < len(p.running) && p.running[minGe] == path[:i] {
-					return p.running[minGe]
+			// Find if there's any P in 'running' such that P is a prefix of 'path'.
+			for i := len(path) - 1; i > 0 && lub >= 0; i-- {
+				// Loop invariants:
+				// 1. lub is the smallest entry in running that is >= path[:i+1] (note
+				//    loop starts at len-1). [2]
+				// 2. path[:i] and [0, lub) are non-empty
+				lub = sort.SearchStrings(p.running[:lub], path[:i])
+				if lub < len(p.running) && p.running[lub] == path[:i] {
+					return p.running[lub]
 				}
 			}
 			return ""
@@ -101,10 +95,10 @@ func (p *pathlock) start(path string) (retErr error) {
 			break
 		}
 	}
-	minGe := sort.SearchStrings(p.running, path)
+	lub := sort.SearchStrings(p.running, path)
 	p.running = append(p.running, "")
-	copy(p.running[minGe+1:], p.running[minGe:]) // OK even if minGe == len(running)
-	p.running[minGe] = path
+	copy(p.running[lub+1:], p.running[lub:]) // OK even if lub == len(running)-1
+	p.running[lub] = path
 	return nil
 }
 
@@ -121,11 +115,11 @@ func (p *pathlock) finish(path string) (retErr error) {
 	p.Lock()
 	defer p.Unlock()
 	// remove 'path' from 'running
-	minGe := sort.SearchStrings(p.running, path)
-	if minGe == len(p.running) || p.running[minGe] != path {
+	lub := sort.SearchStrings(p.running, path)
+	if lub == len(p.running) || p.running[lub] != path {
 		return errors.Errorf("no path lock found for PutFile operation on %q", path)
 	}
-	copy(p.running[minGe:], p.running[minGe+1:]) // OK even if (minGe+1) == len(running)
+	copy(p.running[lub:], p.running[lub+1:]) // OK even if (lub+1) == len(running)
 	p.running = p.running[:len(p.running)-1]
 
 	// Wake up any blocked threads
@@ -140,17 +134,24 @@ func (p *pathlock) finish(path string) (retErr error) {
 	return nil
 }
 
-// [1] Proof:
-// By assumption, we know I = { i ∈ ℕ | path is a prefix of running[i] }
-// is nonempty. Let i be the smallest element of I (well-ordering).
+// [1] lub ("least upper bound") is the smallest entry in 'running' that is
+// greater than or equal to 'path'. This code depends on the fact that if there
+// is a P in 'running' such that 'path' is a prefix of P, then 'path' must be a
+// prefix of 'running[lub]'. Proof:
 //
-// Because path is a prefix of running[i], we know path <= running[i], therefore
-// minGe <= i (by the definition of minGe). Then, because
-//                   path <= running[minGe] <= running[i]
-// path must be a prefix of running[minGe] as well.
+// Let i be the smallest int such that 'path' is a prefix of running[i]. Then,
+// we know 'path' <= running[i] (b/c it's a prefix), and therefore lub <= i (by
+// the definition of lub). Thus, path <= running[lub] <= running[i] and because
+// 'path' is a prefix of running[i], 'path' must be a prefix of running[lub] as
+// well (if any letter was different, it would place running[lub] either before
+// 'path' or after 'running[i]').
 //
-// (if there were some j s.t.  j < len(path) && path[j] != running[minGe], then
-// let j be minimal (w.o.) and we have either 1) path[j] < running[minGe][j]
-// --> running[i][j] < running[minGe][j] --> running[i] < running[minGe], a
-// contradition. Or 2) path[j] > running[minGe][j] --> path > running[minGe], a
-// contradiction)
+// [2] To see that the loop invariant is maintained, we need to show that
+// SearchStrings(running[:lub], path[:i]) yields the least upper bound of
+// path[:i] in running. To see that, observe:
+//   path[:i] < path[:i+1] <= running[lub]
+//   --> SearchStrings(running[:lub], path[:i])
+//         = 1. lub of path[:i] in running[:lub], if any values in running[:lub]
+//              are >= path[:i], or
+//           2. lub, if everything in running[:lub] is < path[:i]
+//         = smallest value in 'running' that is >= path[:i]

--- a/src/server/pfs/server/pathlock.go
+++ b/src/server/pfs/server/pathlock.go
@@ -1,0 +1,156 @@
+package server
+
+import (
+	gopath "path"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/pachyderm/pachyderm/src/client/limit"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+)
+
+// pathlock is used to prevent two PutFile calls that affect the same path (e.g.
+// Put("/foo") and Delete("/")) from racing. The rule is: before running any
+// PutFile operation B, make sure there is no running PutFile operation A such
+// that B.Path is a prefix of A.Path or vice-versa.
+//
+// Note that pathlock can only block one goro at a time. This works for PutFile
+// (where requests are arriving in a stream, and the whole stream stream can be
+// paused until conflicting paths are finished), but you could not use this to
+// schedule N goroutines and have each independently call start() (i.e. many
+// goros blocking at once). Not only would it be complicated to track which
+// goros are blocked on what, this use-case would introduce complicated
+// fairness/livelock issues that pathlock doesn't attempt to solve.
+type pathlock struct {
+	sync.Mutex
+	sync.Cond
+
+	// running contains the set of paths being written to by some in-progress
+	// PutFile operation (we use a set of sorted strings rather than a map so that
+	// we can quickly find prefix collisions)
+	running []string
+
+	// blocking is an in-progress PutFile operation that is blocking the next
+	// PutFile operation in the stream. Before exiting, each putFile goro checks
+	// if it's the blocking operation, and if so, signals to the blocked goro that
+	// it may proceed
+	blocking string
+
+	// This is used to limit the total concurrency of putFile request. It's
+	// included in pathlock as a convenience, as everywhere that putFile calls
+	// pathlock.start(), it would also call putFileLimiter.Acquire() (and likewise
+	// for pathlock.finish() and putFileLimiter.Release). This is unset for tests.
+	putFileLimiter limit.ConcurrencyLimiter
+}
+
+func newPathLock(putFileLimiter limit.ConcurrencyLimiter) *pathlock {
+	p := pathlock{
+		putFileLimiter: putFileLimiter,
+	}
+	p.Cond.L = &p.Mutex
+	return &p
+}
+
+func (p *pathlock) start(path string) (retErr error) {
+	// add leading slash, so '/' collides w/ everything.
+	// TODO(msteffen): would paths.go (in this package) or s/s/pkg/paths be
+	// better? The latter doesn't exist in 1.11.x, and the former adds a leading
+	// slash but doesn't clean the path.
+	path = gopath.Clean("/" + path)
+	defer func() {
+		if retErr == nil && p.putFileLimiter != nil {
+			p.putFileLimiter.Acquire()
+		}
+	}()
+	p.Lock()
+	defer p.Unlock()
+	for { // Check block condition in loop--see sync.Cond documentation
+		if blocking := func() string {
+			// Find if there's any path P in 'running' such that 'path' is a prefix of
+			// P
+			//
+			// minGe is the smallest entry in 'running' that is greater than or equal
+			// to 'path'. If there is a P in running such that 'path' is a prefix of
+			// P, then 'path' must be a prefix of 'running[minGe]'[1]
+			minGe := sort.SearchStrings(p.running, path)
+			if minGe < len(p.running) && strings.HasPrefix(p.running[minGe], path) {
+				return p.running[minGe]
+			}
+			// Find if there's any path P in 'running' such that P is a prefix of
+			// 'path'. Do this by checking if every prefix of 'path' is in 'running'.
+			for i := len(path) - 1; i > 0 && minGe > 0; i-- {
+				// Loop invariant: minGe is the smallest entry in running that is >=
+				//   path[:i+1] (note loop starts at len-1).
+				//   path[:i] < path[:i+1] --> path[:i] must be in [0,minGe)
+				// Loop invariant: path[:i] and [0, minGe) are non-empty
+				minGe = sort.SearchStrings(p.running[:minGe], path[:i])
+				if minGe < len(p.running) && p.running[minGe] == path[:i] {
+					return p.running[minGe]
+				}
+			}
+			return ""
+		}(); blocking != "" {
+			if p.blocking != "" {
+				return errors.Errorf("Internal error: PutFile operation on %q cannot block on %q as an operation is already waiting on %q",
+					path, blocking, p.blocking)
+			}
+			p.blocking = blocking
+			p.Wait() // yield p
+		} else {
+			break
+		}
+	}
+	minGe := sort.SearchStrings(p.running, path)
+	p.running = append(p.running, "")
+	copy(p.running[minGe+1:], p.running[minGe:]) // OK even if minGe == len(running)
+	p.running[minGe] = path
+	return nil
+}
+
+func (p *pathlock) finish(path string) (retErr error) {
+	// add leading slash, so '/' collides w/ everything (see start())
+	path = gopath.Clean("/" + path)
+	defer func() {
+		if p.putFileLimiter != nil {
+			// Call Release() even if finish() returns an error, as the semaphore is
+			// currently locked by the caller.
+			p.putFileLimiter.Release()
+		}
+	}()
+	p.Lock()
+	defer p.Unlock()
+	// remove 'path' from 'running
+	minGe := sort.SearchStrings(p.running, path)
+	if minGe == len(p.running) || p.running[minGe] != path {
+		return errors.Errorf("no path lock found for PutFile operation on %q", path)
+	}
+	copy(p.running[minGe:], p.running[minGe+1:]) // OK even if (minGe+1) == len(running)
+	p.running = p.running[:len(p.running)-1]
+
+	// Wake up any blocked threads
+	if p.blocking == path {
+		p.blocking = "" // reset block
+		// Both Signal() and Broadcast() *should* be safe, b/c there should only
+		// ever be one blocked goro, but I'm always a little worried about having
+		// introduced some subtle deadlock, and because and start() wraps p.Wait()
+		// in a loop anyway, Broadcast() seems like the safer option
+		p.Broadcast()
+	}
+	return nil
+}
+
+// [1] Proof:
+// By assumption, we know I = { i ∈ ℕ | path is a prefix of running[i] }
+// is nonempty. Let i be the smallest element of I (well-ordering).
+//
+// Because path is a prefix of running[i], we know path <= running[i], therefore
+// minGe <= i (by the definition of minGe). Then, because
+//                   path <= running[minGe] <= running[i]
+// path must be a prefix of running[minGe] as well.
+//
+// (if there were some j s.t.  j < len(path) && path[j] != running[minGe], then
+// let j be minimal (w.o.) and we have either 1) path[j] < running[minGe][j]
+// --> running[i][j] < running[minGe][j] --> running[i] < running[minGe], a
+// contradition. Or 2) path[j] > running[minGe][j] --> path > running[minGe], a
+// contradiction)

--- a/src/server/pfs/server/pathlock.go
+++ b/src/server/pfs/server/pathlock.go
@@ -67,12 +67,12 @@ func (p *pathlock) start(path string) (retErr error) {
 	defer p.Unlock()
 	for { // Check block condition in a loop--see sync.Cond documentation
 		if blocking := func() string {
-			// Find if there's any P in 'running' such that 'path' is a prefix of P[1]
+			// Find any P in 'running' such that 'path' is a prefix of P. [1]
 			lub := sort.SearchStrings(p.running, path)
 			if lub < len(p.running) && strings.HasPrefix(p.running[lub], path) {
 				return p.running[lub]
 			}
-			// Find if there's any P in 'running' such that P is a prefix of 'path'.
+			// Find any P in 'running' such that P is a prefix of 'path'.
 			for i := len(path) - 1; i > 0 && lub >= 0; i-- {
 				// Loop invariants:
 				// 1. lub is the smallest entry in running that is >= path[:i+1] (note
@@ -146,10 +146,11 @@ func (p *pathlock) finish(path string) (retErr error) {
 // well (if any letter was different, it would place running[lub] either before
 // 'path' or after 'running[i]').
 //
-// [2] To see that the loop invariant is maintained, we need to show that
-// SearchStrings(running[:lub], path[:i]) yields the least upper bound of
-// path[:i] in running. To see that, observe:
-//   path[:i] < path[:i+1] <= running[lub]
+// [2] To see that the loop invariant 'path[:i+1] <= running[lub]' is
+// maintained, we need to show that
+//   lub' = SearchStrings(running[:lub], path[:i]) = lub of path[:i] in running
+// yields the least upper bound of path[:i] in running. To see that, observe:
+//   path[:i] < path[:i+1] <= running[lub] <= running[lub+1:]...
 //   --> SearchStrings(running[:lub], path[:i])
 //         = 1. lub of path[:i] in running[:lub], if any values in running[:lub]
 //              are >= path[:i], or

--- a/src/server/pfs/server/pathlock_test.go
+++ b/src/server/pfs/server/pathlock_test.go
@@ -1,0 +1,162 @@
+package server
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/pachyderm/pachyderm/src/client/pkg/require"
+)
+
+// TestPathLockSomeCollisions starts a bunch of threads for paths, some of which
+// overlap and some of which don't. This is the main way that we expect PutFile
+// to work.
+func TestPathLockSomeCollisions(t *testing.T) {
+	var (
+		seed       = time.Now().UnixNano()
+		r          = rand.New(rand.NewSource(seed))
+		running    = make(map[string]bool) // paths held by currently running ops
+		runningMu  sync.Mutex              // guards 'running'
+		runningStr = func() []string {     // print 'running' as a list, not a map
+			strs := make([]string, 0, len(running))
+			for k := range running {
+				strs = append(strs, k)
+			}
+			return strs
+		}
+		start, end   = true, false                  // syntax sugar for checkRunning
+		maxRunning   = 0                            // count the max # of running ops
+		checkRunning = func(op bool, path string) { // check 'running' for prefix collisions
+			runningMu.Lock()
+			defer runningMu.Unlock()
+			if len(running) > maxRunning {
+				maxRunning = len(running)
+			}
+			for p := range running {
+				if op == end && path == p {
+					continue // we expect path itself to be in running if we're removing it
+				}
+				// no prefix of 'path' is running (as this isn't the main test goro, use
+				// Errorf instead of require.False so that execution isn't halted on
+				// failure, preventing deadlock)
+				if strings.HasPrefix(path, p) {
+					t.Errorf("(seed: %d) path %q has prefix %s ∈ %v", seed, path, p, runningStr())
+				}
+				// 'path' is not a prefix of any running op
+				if strings.HasPrefix(p, path) {
+					t.Errorf("(seed: %d) path %q is a prefix of %s ∈ %v", seed, path, p, runningStr())
+				}
+			}
+			if op == start {
+				running[path] = true
+			} else {
+				delete(running, path)
+			}
+		}
+		p  = newPathLock() // the data structure under test
+		eg errgroup.Group  // for running the subprocesses
+	)
+	// Iterate through [0,1000) in a random order, represented as binary strings.
+	// This induces regular collisions (these strings aren't fixed-width, so small
+	// numbers' representations are often prefixes of larger numbers), but also
+	// allows for some parallelism.
+	//
+	// This test uses 'maxRunning' to check that some parallelism occurs, and the
+	// prevention of collisions can be verified by removing 'p.start(path)' and
+	// 'p.finish(path)' below, which causes the collision checks in 'checkRunning'
+	// to fail.
+	for _, i := range r.Perm(1000) {
+		// small nums are prefixes of bigger nums. Iterate all nums but jump around
+		// to induce collisions
+		path := fmt.Sprintf("%b", i)
+		require.NoError(t, p.start(path))
+		eg.Go(func() error {
+			defer func() { require.NoError(t, p.finish(path)) }()
+			checkRunning(start, path)
+			// force later prefixes to run ~first
+			time.Sleep(time.Duration(100-i) * time.Millisecond)
+			checkRunning(end, path)
+			return nil
+		})
+	}
+	require.NoError(t, eg.Wait())
+	// want at least some concurrency in this test
+	require.True(t, maxRunning > 1, maxRunning)
+}
+
+// TestPathLockAllCollisions is the strictest test, and starts several
+// goroutines where every goroutine is given a path that extends the previous
+// goroutines' paths. There should be no overlap, and later goroutines should
+// wait on earlier goroutines
+func TestPathLockAllCollisions(t *testing.T) {
+	p := newPathLock()
+	// Lock every prefix of 'paths' and make sure that they're processed in order
+	paths := strings.Repeat("a", 100)
+	ch := make(chan string)
+	// write all paths in a separate goro so the main goro can read from 'ch'
+	go func() {
+		var eg errgroup.Group
+		for i := range paths {
+			path := paths[:i]
+			require.NoError(t, p.start(path))
+			eg.Go(func() error {
+				defer func() { require.NoError(t, p.finish(path)) }()
+				// force later prefixes to run first
+				time.Sleep(time.Duration(100-i) * time.Millisecond)
+				ch <- path
+				return nil
+			})
+		}
+		require.NoError(t, eg.Wait())
+		close(ch)
+	}()
+
+	var prev string
+	for p := range ch {
+		if prev != "" {
+			require.True(t, len(prev) < len(p), "expected %q to be shorter than %q", p, prev)
+		}
+		prev = p
+	}
+}
+
+// TestPathLockAllCollisionsRev is similar to TestPathLockAllCollisions, but it
+// iterates over the colliding paths from longest to shortest, to ensure that
+// pathlock handles that direction as well (blocking paths if they're a prefix
+// of a running operation's path)
+func TestPathLockAllCollisionsRev(t *testing.T) {
+	p := newPathLock()
+	// Lock every prefix of 'paths' and make sure that they're processed in order
+	paths := strings.Repeat("a", 100)
+	ch := make(chan string)
+	// write all paths in a separate goro so the main goro can read from 'ch'
+	go func() {
+		var eg errgroup.Group
+		for i := range paths {
+			path := paths[:len(paths)-1-i]
+			require.NoError(t, p.start(path))
+			eg.Go(func() error {
+				defer func() { require.NoError(t, p.finish(path)) }()
+				// force later prefixes to run first
+				time.Sleep(time.Duration(100-i) * time.Millisecond)
+				ch <- path
+				return nil
+			})
+		}
+		require.NoError(t, eg.Wait())
+		close(ch)
+	}()
+
+	var prev string
+	for p := range ch {
+		if prev != "" {
+			require.True(t, len(prev) > len(p), "expected %q to be shorter than %q", p, prev)
+		}
+		prev = p
+	}
+}

--- a/src/server/pfs/server/pathlock_test.go
+++ b/src/server/pfs/server/pathlock_test.go
@@ -58,8 +58,8 @@ func TestPathLockSomeCollisions(t *testing.T) {
 				delete(running, path)
 			}
 		}
-		p  = newPathLock() // the data structure under test
-		eg errgroup.Group  // for running the subprocesses
+		p  = newPathLock(nil) // the data structure under test
+		eg errgroup.Group     // for running the subprocesses
 	)
 	// Iterate through [0,1000) in a random order, represented as binary strings.
 	// This induces regular collisions (these strings aren't fixed-width, so small
@@ -94,7 +94,7 @@ func TestPathLockSomeCollisions(t *testing.T) {
 // goroutines' paths. There should be no overlap, and later goroutines should
 // wait on earlier goroutines
 func TestPathLockAllCollisions(t *testing.T) {
-	p := newPathLock()
+	p := newPathLock(nil)
 	// Lock every prefix of 'paths' and make sure that they're processed in order
 	paths := strings.Repeat("a", 100)
 	ch := make(chan string)
@@ -130,7 +130,7 @@ func TestPathLockAllCollisions(t *testing.T) {
 // pathlock handles that direction as well (blocking paths if they're a prefix
 // of a running operation's path)
 func TestPathLockAllCollisionsRev(t *testing.T) {
-	p := newPathLock()
+	p := newPathLock(nil)
 	// Lock every prefix of 'paths' and make sure that they're processed in order
 	paths := strings.Repeat("a", 100)
 	ch := make(chan string)

--- a/src/server/pfs/server/testing/server_test.go
+++ b/src/server/pfs/server/testing/server_test.go
@@ -578,6 +578,43 @@ func TestPutFileDirectoryTraversal(t *testing.T) {
 	require.NoError(t, err)
 }
 
+// TestPutFileOverlappingPaths tests the fix for
+// https://github.com/pachyderm/pachyderm/issues/5345.
+// However, I can't get the test to fail without adding a sleep to
+// forEachPutFile in driver.go that induces the race. Setting GOMAXPROCS to 1
+// and 100 doesn't seem to help. In practice, that means we're still relying on
+// TestPipelineBuildLifecycle and pachyderm-in-Minikube to expose the race
+// (Minikube seems to induce more races), but maybe this test will be useful in
+// conjunction with e.g. some kind of race detector.
+// TODO(msteffen): Get this test to fail reliably in the presence of the race.
+func TestPutFileOverlappingPaths(t *testing.T) {
+	t.Parallel()
+	files := []string{"/a", "/b", "/c"}
+	contents := []string{"foo", "bar", "baz"}
+	err := testpachd.WithRealEnv(func(env *testpachd.RealEnv) error {
+		c := env.PachClient
+		c.CreateRepo("test")
+		for i := 0; i < 100; i++ {
+			pfc, err := c.NewPutFileClient()
+			require.NoError(t, err)
+			require.NoError(t, pfc.DeleteFile("test", "master", "/"))
+			for i, f := range files {
+				_, err := pfc.PutFile("test", "master", f, strings.NewReader(contents[i]))
+				require.NoError(t, err)
+			}
+			require.NoError(t, pfc.Close())
+			var actualFiles []string
+			c.ListFileF("test", "master", "/", 0, func(f *pfs.FileInfo) error {
+				actualFiles = append(actualFiles, f.File.Path)
+				return nil
+			})
+			require.ElementsEqual(t, files, actualFiles)
+		}
+		return nil
+	})
+	require.NoError(t, err)
+}
+
 func TestCreateInvalidBranchName(t *testing.T) {
 	t.Parallel()
 	err := testpachd.WithRealEnv(func(env *testpachd.RealEnv) error {
@@ -6422,11 +6459,16 @@ func TestAtomicHistory(t *testing.T) {
 
 type SlowReader struct {
 	underlying io.Reader
+	delay      time.Duration
 }
 
 func (r *SlowReader) Read(p []byte) (n int, err error) {
 	n, err = r.underlying.Read(p)
-	time.Sleep(1 * time.Millisecond)
+	if r.delay == 0 {
+		time.Sleep(1 * time.Millisecond)
+	} else {
+		time.Sleep(r.delay)
+	}
 	return
 }
 

--- a/src/server/pfs/server/testing/server_test.go
+++ b/src/server/pfs/server/testing/server_test.go
@@ -581,8 +581,9 @@ func TestPutFileDirectoryTraversal(t *testing.T) {
 // TestPutFileOverlappingPaths tests the fix for
 // https://github.com/pachyderm/pachyderm/issues/5345.
 // However, I can't get the test to fail without adding a sleep to
-// forEachPutFile in driver.go that induces the race. Setting GOMAXPROCS to 1
-// and 100 doesn't seem to help. In practice, that means we're still relying on
+// forEachPutFile in driver.go that induces the race (at:
+// `if req.Delete {...eg.Go(/*here*/...)}`). Setting GOMAXPROCS to 1 and 100
+// doesn't seem to help. In practice, that means we're still relying on
 // TestPipelineBuildLifecycle and pachyderm-in-Minikube to expose the race
 // (Minikube seems to induce more races), but maybe this test will be useful in
 // conjunction with e.g. some kind of race detector.

--- a/src/server/pps/cmds/cmds_test.go
+++ b/src/server/pps/cmds/cmds_test.go
@@ -28,7 +28,6 @@ import (
 	"bytes"
 	"fmt"
 	"os"
-	"os/exec"
 	"testing"
 
 	"github.com/pachyderm/pachyderm/src/client/pkg/require"
@@ -940,14 +939,14 @@ func runPipelineWithImageGetStderr(t *testing.T, image string) (string, error) {
 		pachctl put file -r in@master:/ -f ../../../../etc/testing/pipeline-build/input
 	`).Run())
 
-	cmd := exec.Command("/bin/bash", "-c", fmt.Sprintf(`
+	cmd := tu.BashCmd(`
 		pachctl create pipeline <<EOF
 		  {
 		    "pipeline": {
 		      "name": "first"
 		    },
 		    "transform": {
-		      "image": "%s"
+		      "image": "{{.image}}"
 		    },
 		    "input": {
 		      "pfs": {
@@ -957,7 +956,7 @@ func runPipelineWithImageGetStderr(t *testing.T, image string) (string, error) {
 		    }
 		  }
 		EOF
-	`, image))
+	`, "image", image)
 	buf := &bytes.Buffer{}
 	cmd.Stderr = buf
 	// cmd.Stdout = os.Stdout // uncomment for debugging
@@ -1009,7 +1008,7 @@ func TestWarningLatestTag(t *testing.T) {
 	}
 	// should emit a warning because user specified latest tag on docker image
 	stderr, err := runPipelineWithImageGetStderr(t, "ubuntu:latest")
-	require.NoError(t, err)
+	require.NoError(t, err, "%v", err)
 	require.Matches(t, "WARNING", stderr)
 }
 

--- a/src/server/pps/cmds/cmds_test.go
+++ b/src/server/pps/cmds/cmds_test.go
@@ -610,16 +610,16 @@ func TestEditPipeline(t *testing.T) {
 }
 
 func TestPipelineBuildLifecyclePython(t *testing.T) {
-	testPipelineBuildLifecycle(t, "python", "python")
+	pipeline := testPipelineBuildLifecycle(t, "python", "python")
 
 	// the python example also contains a `.pachignore`, so we can verify it's
 	// intended behavior here
 	require.NoError(t, tu.BashCmd(`
-		pachctl get file test-pipeline-build_build@source:/.pachignore
-	`).Run())
+		pachctl get file {{.pipeline}}_build@source:/.pachignore
+	`, "pipeline", pipeline).Run())
 	require.YesError(t, tu.BashCmd(`
-		pachctl get file test-pipeline-build_build@source:/foo.txt
-	`).Run())
+		pachctl get file {{.pipeline}}_build@source:/foo.txt
+	`, "pipeline", pipeline).Run())
 }
 
 func TestPipelineBuildLifecyclePythonNoDeps(t *testing.T) {
@@ -630,7 +630,7 @@ func TestPipelineBuildLifecycleGo(t *testing.T) {
 	testPipelineBuildLifecycle(t, "go", "go")
 }
 
-func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
+func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 	t.Helper()
 	if testing.Short() {
 		t.Skip("Skipping integration tests in short mode")
@@ -643,14 +643,17 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		pachctl put file -r in@master:/ -f ../../../../etc/testing/pipeline-build/input
 	`).Run())
 
+	// give pipeline a unique name to work around
+	// github.com/kubernetes/kubernetes/issues/82130
+	pipeline := tu.UniqueString("test-pipeline-build")
 	spec := fmt.Sprintf(`
 		{
 		  "pipeline": {
-		    "name": "test-pipeline-build"
+		    "name": %q
 		  },
 		  "transform": {
 		    "build": {
-		      "language": "%s"
+		      "language": %q
 		    }
 		  },
 		  "input": {
@@ -660,7 +663,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		    }
 		  }
 		}
-	`, lang)
+	`, pipeline, lang)
 
 	// test a barebones pipeline with a build spec and verify results
 	require.NoError(t, tu.BashCmd(`
@@ -673,7 +676,10 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		"dir", dir,
 		"spec", spec,
 	).Run())
-	verifyPipelineBuildOutput(t, "0")
+	require.YesError(t, tu.BashCmd(fmt.Sprintf(`
+		pachctl list pipeline --state failure | match %s
+	`, pipeline)).Run())
+	verifyPipelineBuildOutput(t, pipeline, "0")
 
 	// update the barebones pipeline and verify results
 	require.NoError(t, tu.BashCmd(`
@@ -686,13 +692,13 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		"dir", dir,
 		"spec", spec,
 	).Run())
-	verifyPipelineBuildOutput(t, "0")
+	verifyPipelineBuildOutput(t, pipeline, "0")
 
 	// update the pipeline with a custom cmd and verify results
 	spec = fmt.Sprintf(`
 		{
 		  "pipeline": {
-		    "name": "test-pipeline-build"
+		    "name": %q
 		  },
 		  "transform": {
 		    "cmd": [
@@ -701,7 +707,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		      "_"
 		    ],
 		    "build": {
-		      "language": "%s",
+		      "language": %q,
 		      "path": "."
 		    }
 		  },
@@ -712,7 +718,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		    }
 		  }
 		}
-	`, lang)
+	`, pipeline, lang)
 
 	require.NoError(t, tu.BashCmd(`
 		cd ../../../../etc/testing/pipeline-build/{{.dir}}
@@ -724,18 +730,20 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) {
 		"dir", dir,
 		"spec", spec,
 	).Run())
-	verifyPipelineBuildOutput(t, "_")
+	verifyPipelineBuildOutput(t, pipeline, "_")
+	return pipeline
 }
 
-func verifyPipelineBuildOutput(t *testing.T, prefix string) {
+func verifyPipelineBuildOutput(t *testing.T, pipeline, prefix string) {
 	t.Helper()
 
 	require.NoError(t, tu.BashCmd(`
 		pachctl flush commit test-pipeline-build@master
-		pachctl get file test-pipeline-build@master:/1.txt | match {{.prefix}}{{.prefix}}{{.prefix}}1
-		pachctl get file test-pipeline-build@master:/11.txt | match {{.prefix}}{{.prefix}}11
-		pachctl get file test-pipeline-build@master:/111.txt | match {{.prefix}}111
+		pachctl get file {{.pipeline}}@master:/1.txt | match {{.prefix}}{{.prefix}}{{.prefix}}1
+		pachctl get file {{.pipeline}}@master:/11.txt | match {{.prefix}}{{.prefix}}11
+		pachctl get file {{.pipeline}}@master:/111.txt | match {{.prefix}}111
 		`,
+		"pipeline", pipeline,
 		"prefix", prefix,
 	).Run())
 }
@@ -976,7 +984,7 @@ func TestPipelineBuildRunCron(t *testing.T) {
 				"name": "tick",
 				"spec": "*/1 * * * *",
 				"overwrite": true
-			  }  
+			  }
 			},
 			"transform": {
 			  "build":{

--- a/src/server/pps/cmds/cmds_test.go
+++ b/src/server/pps/cmds/cmds_test.go
@@ -671,7 +671,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 		pachctl create pipeline <<EOF
 			{{.spec}}
 		EOF
-		pachctl flush commit test-pipeline-build@master
+		pachctl flush commit in@master
 		`,
 		"dir", dir,
 		"spec", spec,
@@ -687,7 +687,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 		pachctl update pipeline <<EOF
 			{{.spec}}
 		EOF
-		pachctl flush commit test-pipeline-build@master
+		pachctl flush commit in@master
 		`,
 		"dir", dir,
 		"spec", spec,
@@ -725,7 +725,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 		pachctl update pipeline --reprocess <<EOF
 			{{.spec}}
 		EOF
-		pachctl flush commit test-pipeline-build@master
+		pachctl flush commit in@master
 		`,
 		"dir", dir,
 		"spec", spec,
@@ -738,7 +738,7 @@ func verifyPipelineBuildOutput(t *testing.T, pipeline, prefix string) {
 	t.Helper()
 
 	require.NoError(t, tu.BashCmd(`
-		pachctl flush commit test-pipeline-build@master
+		pachctl flush commit in@master
 		pachctl get file {{.pipeline}}@master:/1.txt | match {{.prefix}}{{.prefix}}{{.prefix}}1
 		pachctl get file {{.pipeline}}@master:/11.txt | match {{.prefix}}{{.prefix}}11
 		pachctl get file {{.pipeline}}@master:/111.txt | match {{.prefix}}111

--- a/src/server/pps/cmds/cmds_test.go
+++ b/src/server/pps/cmds/cmds_test.go
@@ -99,19 +99,19 @@ func TestRawFullPipelineInfo(t *testing.T) {
 		pachctl create repo data
 		pachctl put file data@master:/file <<<"This is a test"
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {"name": "{{.pipeline}}"},
-			  "input": {
-			    "pfs": {
-			      "glob": "/*",
-			      "repo": "data"
-			    }
-			  },
-			  "transform": {
-			    "cmd": ["bash"],
-			    "stdin": ["cp /pfs/data/file /pfs/out"]
-			  }
-			}
+		  {
+		    "pipeline": {"name": "{{.pipeline}}"},
+		    "input": {
+		      "pfs": {
+		        "glob": "/*",
+		        "repo": "data"
+		      }
+		    },
+		    "transform": {
+		      "cmd": ["bash"],
+		      "stdin": ["cp /pfs/data/file /pfs/out"]
+		    }
+		  }
 		EOF
 		`,
 		"pipeline", tu.UniqueString("p-")).Run())
@@ -208,9 +208,9 @@ func TestJSONStringifiedNumbers(t *testing.T) {
 		      "repo": "input"
 		    }
 		  },
-			"parallelism_spec": {
-				"constant": "1"
-			},
+		  "parallelism_spec": {
+		    "constant": "1"
+		  },
 		  "transform": {
 		    "cmd": [ "/bin/bash" ],
 		    "stdin": [
@@ -389,22 +389,22 @@ func TestListPipelineFilter(t *testing.T) {
 		pachctl create pipeline -f - <<EOF
 		{
 		"pipeline": {
-			"name": "first"
+		  "name": "first"
 		},
 		"input": {
-			"pfs": {
-			"glob": "/*",
-			"repo": "input"
-			}
+		  "pfs": {
+		    "glob": "/*",
+		    "repo": "input"
+		  }
 		},
-			"parallelism_spec": {
-				"constant": "1"
-			},
+		  "parallelism_spec": {
+		    "constant": "1"
+		  },
 		"transform": {
-			"cmd": [ "/bin/bash" ],
-			"stdin": [
-			"cp /pfs/input/* /pfs/out"
-			]
+		  "cmd": [ "/bin/bash" ],
+		  "stdin": [
+		    "cp /pfs/input/* /pfs/out"
+		  ]
 		}
 		}
 		EOF
@@ -669,7 +669,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 	require.NoError(t, tu.BashCmd(`
 		cd ../../../../etc/testing/pipeline-build/{{.dir}}
 		pachctl create pipeline <<EOF
-			{{.spec}}
+		{{.spec}}
 		EOF
 		pachctl flush commit in@master
 		`,
@@ -685,7 +685,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 	require.NoError(t, tu.BashCmd(`
 		cd ../../../../etc/testing/pipeline-build/{{.dir}}
 		pachctl update pipeline <<EOF
-			{{.spec}}
+		{{.spec}}
 		EOF
 		pachctl flush commit in@master
 		`,
@@ -723,7 +723,7 @@ func testPipelineBuildLifecycle(t *testing.T, lang, dir string) string {
 	require.NoError(t, tu.BashCmd(`
 		cd ../../../../etc/testing/pipeline-build/{{.dir}}
 		pachctl update pipeline --reprocess <<EOF
-			{{.spec}}
+		{{.spec}}
 		EOF
 		pachctl flush commit in@master
 		`,
@@ -755,17 +755,17 @@ func TestMissingPipeline(t *testing.T) {
 	// should fail because there's no pipeline object in the spec
 	require.YesError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
-			{
-			  "transform": {
-			    "image": "ubuntu"
-			  },
-			  "input": {
-			    "pfs": {
-			      "repo": "in",
-			      "glob": "/*"
-			    }
-			  }
-			}
+		  {
+		    "transform": {
+		      "image": "ubuntu"
+		    },
+		    "input": {
+		      "pfs": {
+		        "repo": "in",
+		        "glob": "/*"
+		      }
+		    }
+		  }
 		EOF
 	`).Run())
 }
@@ -777,18 +777,18 @@ func TestUnnamedPipeline(t *testing.T) {
 	// should fail because there's no pipeline name
 	require.YesError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {},
-			  "transform": {
-			    "image": "ubuntu"
-			  },
-			  "input": {
-			    "pfs": {
-			      "repo": "in",
-			      "glob": "/*"
-			    }
-			  }
-			}
+		  {
+		    "pipeline": {},
+		    "transform": {
+		      "image": "ubuntu"
+		    },
+		    "input": {
+		      "pfs": {
+		        "repo": "in",
+		        "glob": "/*"
+		      }
+		    }
+		  }
 		EOF
 	`).Run())
 }
@@ -800,17 +800,17 @@ func TestPipelineBuildSpout(t *testing.T) {
 	// should fail because pipeline build can't be used w/ spouts
 	require.YesError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {
-			    "name": "test"
-			  },
-			  "transform": {
-			    "build": {
-			      "language": "go"
-			    }
-			  },
-			  "spout": {}
-			}
+		  {
+		    "pipeline": {
+		      "name": "test"
+		    },
+		    "transform": {
+		      "build": {
+		        "language": "go"
+		      }
+		    },
+		    "spout": {}
+		  }
 		EOF
 	`).Run())
 }
@@ -822,16 +822,16 @@ func TestPipelineBuildMissingInput(t *testing.T) {
 	// should fail because pipeline build can't be used w/ spouts
 	require.YesError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {
-			    "name": "test"
-			  },
-			  "transform": {
-			    "build": {
-			      "language": "go"
-			    }
-			  }
-			}
+		  {
+		    "pipeline": {
+		      "name": "test"
+		    },
+		    "transform": {
+		      "build": {
+		        "language": "go"
+		      }
+		    }
+		  }
 		EOF
 	`).Run())
 }
@@ -844,44 +844,44 @@ func TestPipelineBuildBadInput(t *testing.T) {
 	// using pipeline builds
 	require.YesError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {
-			    "name": "test"
-			  },
-			  "transform": {
-			    "build": {
-			      "language": "go"
-			    }
-			  },
-			  "input": {
-			    "pfs": {
-			      "name": "build",
-			      "repo": "in",
-			      "glob": "/*"
-			    }
-			  }
-			}
+		  {
+		    "pipeline": {
+		      "name": "test"
+		    },
+		    "transform": {
+		      "build": {
+		        "language": "go"
+		      }
+		    },
+		    "input": {
+		      "pfs": {
+		        "name": "build",
+		        "repo": "in",
+		        "glob": "/*"
+		      }
+		    }
+		  }
 		EOF
 	`).Run())
 	require.YesError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {
-			    "name": "test"
-			  },
-			  "transform": {
-			    "build": {
-			      "language": "go"
-			    }
-			  },
-			  "input": {
-			    "pfs": {
-			      "name": "source",
-			      "repo": "in",
-			      "glob": "/*"
-			    }
-			  }
-			}
+		  {
+		    "pipeline": {
+		      "name": "test"
+		    },
+		    "transform": {
+		      "build": {
+		        "language": "go"
+		      }
+		    },
+		    "input": {
+		      "pfs": {
+		        "name": "source",
+		        "repo": "in",
+		        "glob": "/*"
+		      }
+		    }
+		  }
 		EOF
 	`).Run())
 }
@@ -942,21 +942,21 @@ func runPipelineWithImageGetStderr(t *testing.T, image string) (string, error) {
 
 	cmd := exec.Command("/bin/bash", "-c", fmt.Sprintf(`
 		pachctl create pipeline <<EOF
-			{
-			  "pipeline": {
-				"name": "first"
-			  },
-			  "transform": {
-				"image": "%s"
-			  },
-			  "input": {
-			    "pfs": {
-			      "repo": "in",
-			      "glob": "/*"
-			    }
-			  }
-			}
-EOF
+		  {
+		    "pipeline": {
+		      "name": "first"
+		    },
+		    "transform": {
+		      "image": "%s"
+		    },
+		    "input": {
+		      "pfs": {
+		        "repo": "in",
+		        "glob": "/*"
+		      }
+		    }
+		  }
+		EOF
 	`, image))
 	buf := &bytes.Buffer{}
 	cmd.Stderr = buf
@@ -976,23 +976,23 @@ func TestPipelineBuildRunCron(t *testing.T) {
 	require.NoError(t, tu.BashCmd(`
 		pachctl create pipeline <<EOF
 		{
-			"pipeline": {
-			  "name": "crontest"
-			},
-			"input": {
-			  "cron": {
-				"name": "tick",
-				"spec": "*/1 * * * *",
-				"overwrite": true
-			  }
-			},
-			"transform": {
-			  "build":{
-				"language": "go"
-			  },
-			  "cmd":["echo", "tick"]
-			},
-			"enable_stats": true
+		  "pipeline": {
+		    "name": "crontest"
+		  },
+		  "input": {
+		    "cron": {
+		      "name": "tick",
+		      "spec": "*/1 * * * *",
+		      "overwrite": true
+		    }
+		  },
+		  "transform": {
+		    "build":{
+		      "language": "go"
+		    },
+		    "cmd":["echo", "tick"]
+		  },
+		  "enable_stats": true
 		  }
 		EOF
 	`).Run())

